src/gui/transferlistsortmodel.h

```diff
@@ -57,5 +59,8 @@ class TransferListSortModel final : public QSortFilterProxyModel
     bool filterAcceptsRow(int sourceRow, const QModelIndex &sourceParent) const override;
     bool matchFilter(int sourceRow, const QModelIndex &sourceParent) const;
 
+    int compare(const QModelIndex &left, const QModelIndex &right) const;
+
     TorrentFilter m_filter;
+    mutable boost::circular_buffer<int> m_sortColumns {boost::circular_buffer<int>::size_type {2}};
```

Comment:
Seems you really need to store one previous sort column. So circular buffer is redundant, isn't it?

Reply:
oof

Reply:
it wouldn't be as clean as I initially thought, I have to maintain an extra flag to differentiate the initial case when there are no secondary criteria for sorting.

Reply:
Isn't `std::optional<int>` enough?

Reply:
> Isn't std::optional<int> enough?

I didn't try implementing but IMO it would still be more complicated because of the initial case when there is no secondary criteria for sorting. It seems a bit difficult to know when to change the secondary criteria without maintaining a list-like structure.

Reply:
```c++
bool TransferListSortModel::lessThan(const QModelIndex &left, const QModelIndex &right) const
{
    Q_ASSERT(left.column() == right.column());

    std::optional<int> prevColumn = m_prevColumn;
    m_prevColumn = left.column();
    const int result = compare(left, right);
    if (result != 0)
        return result < 0;

    if (prevColumn)
        return compare(left.sibling(left.row(), *prevColumn), right.sibling(right.row(), *prevColumn);

    return false; // initial comparison gave equality
}
```

Reply:
Generic algorithm using circular buffer looks confusing. To really understand what's happening (there is buffer of capacity of 2 and it is never changed) someone need to review entire transferlistsortmodel.cpp/.h files.

Reply:
```cpp
bool TransferListSortModel::lessThan(const QModelIndex &left, const QModelIndex &right) const
{
    Q_ASSERT(left.column() == right.column());

    std::optional<int> prevColumn = m_prevColumn;
    m_prevColumn = left.column();
    const int result = compare(left, right);
    if (result != 0)
        return result < 0;

    if (prevColumn)
        return compare(left.sibling(left.row(), *prevColumn), right.sibling(right.row(), *prevColumn);

    return false; // initial comparison gave equality
}
```

This won't work, you always overwrite m_prevColumn, it seems you assumed that lessThan is only called once after sortColumn is changed which isn't the case, it will be called with every item after sortColumn changes or data changes.

Reply:
>This won't work, you always overwrite m_prevColumn, it seems you assumed that lessThan is only called once after sortColumn is changed which isn't the case, it will be called with every item after sortColumn changes or data changes.

Ok, seems you're right. But https://github.com/qbittorrent/qBittorrent/pull/14423/files#r579790552 is still actual. At least you should add a comment in lessThan with an explanation that buffer has constant capacity of 2 so it always contains no more than 2 distinct values.

Reply:
done.

Reply:
Doesn't `mutable boost::circular_buffer<int> m_sortColumns {2};` work?